packages/flutter/lib/src/foundation/annotations.dart

```diff
@@ -21,9 +19,10 @@
 /// ```dart
 /// /// A copper coffee pot, as desired by Ben Turpin.
 /// /// ...documentation...
-/// @Category(<String>['Pots', 'Coffee'])
-/// @Category(<String>['Copper', 'Cookware'])
-/// @DocumentationIcon('https://example.com/images/coffee.png')
+/// @Category('Pots')
```

Comment:
How do Category and Subcategory actually relate? Do "Pots" and "Copper" each have two sub-categories ("Coffee" and "Cookware")?

Reply:
Wouldn't subcategories be category specific? Should Category maybe have a subcategory property?

Reply:
From actual usage in the existing widget catalog, category and subcategory seem to be independent; a widget can exist in a subcategory without being in a category.

https://github.com/flutter/website/blob/master/src/_data/catalog/widgets.json

What I'm trying to do for this PR is to make these annotations better map to the usage in the widget catalog, and then look at landing a PR which can build a machine generated catalog from the package:flutter source.

Once it's machine generated, it'll be easier to evolve the annotations, tool, and generated catalog all as a unit.

Reply:
Can we find a better name for this concept? Looking at the json it's not clear to me when something is a subcategory and when it is a regular category.

Reply:
I looked at the example of the [Align](https://github.com/flutter/website/blob/master/src/_data/catalog/widgets.json#L23) widget in widgets.json, which has a subcategory but not a category. However, in the [actual catalog page](https://flutter.dev/docs/development/ui/widgets/layout), it obviously is under the top category "Layout widgets". I'm not sure where the disconnect is.

Reply:
I _think_ what's going on here is that the sidecar index.json file has some mapping of sub-categories to categories:

https://github.com/flutter/website/blob/master/src/_data/catalog/index.json#L39

So we're able to generate the web pages, even without full category info in the widgets.json (catalog) file.

Reply:
I do see a few widgets that have multiple categories (like `Image`):

```
"categories": [
      "Basics",
      "Assets, Images, and Icons"
    ],
```

and a few (like `GridView`) with multiple subcategories:

```
"subcategories": [
      "Information displays",
      "Multi-child layout widgets"
    ],
```

I don't know if a subcategory always implies a specific category. If it does, then we could re-define the annotation to be `@Category('main category', subcategory: 'a sub-category)`.

And, I don't know how valuable the subcategories are. If they're cross-cutting properties, perhaps we should just call them something like `@ClassificationProperty('foo')` or `@Tag('foo')` or `@ClassificationTag('foo')`.

Reply:
The logic of generating the catalog webpages is specified in [this template file](https://github.com/flutter/website/blob/master/src/_includes/catalogpage.html). IIUC, the template first grabs a category from [index.json](https://github.com/flutter/website/blob/master/src/_data/catalog/index.json) and generate HTML for all the entries in [widget.json](https://github.com/flutter/website/blob/master/src/_data/catalog/widgets.json) that has that category. It then looks at each subcategory, if any, within the category and generate HTML for all the entries in [widget.json](https://github.com/flutter/website/blob/master/src/_data/catalog/widgets.json) that has that subcategory. Given this process, omitting categories for widgets that have a subcategory is actually _necessary_  to avoid generating HTML for those widgets under both the matching categories and the matching subcategories.

That logic is a bit complicated. Would you like to clean it up or just want to match what it is today? Conceptually, i think each widget should have one or more categories it can belong to, and within each category, the widget might be further grouped into a subcategories. I'm imagining something like this: 

```
[
  {'category': 'Layout widgets', 'subcategory': 'Single-child layout widgets'},
  {'category': 'Material Design widgets'},
]

```
I'm not sure how to express the above in annotations.

Reply:
To be more general, is it useful to be able to describe a deeper hierarchy?

Widgets could be tagged with one or more "Category" tags, and then something else (a JSON file somewhere, maybe) would describe what the hierarchy of categories looks like. We could have tooling to make sure that all categories appear in the JSON file (and that all categories in the JSON file have at least one tag in the source).

It would allow us to rearrange the hierarchy of categories without editing all the code, too.

Reply:
Although I don't like that all the data isn't available in the source...  Hmm.

Maybe categories are just lists:
```dart
/// @Category(['Material widgets', 'Layout widgets', 'Single-child layout widgets'])
```

Reply:
Maybe we can re-conceptualize categories and subcategories as paths to find a widget. A widget can have multiple paths for finding it, represented as a list. Each path can include 1+ levels of categories, again represented as a list which reads from left to right. So in the following example, the widget has two paths: 1) Material widgets, and 2) Layout widgets --> Single-child layout widgets.  
```
/// @Category([['Material widgets'], ['Layout widgets', 'Single-child layout widgets']])
```

Reply:
That's basically the same as the original design, it just merges multiple `@Category` annotations into one.

Reply:
Okay, it looks like so. :)

Reply:
I'm slightly lost. Does this describe what we're talking about:

```
@Category(
  'Layout widgets',
  subcategories: ['Information displays', 'Single-child layout widgets'],
)
```

with multiple category annotations allowed per widget.

If so, if there a preferences between the above form, or a `@Category` annotation just taking a list of strings, which represents a single path from a category through to a specific sub-category?

Reply:
We can definitely do validation on the source, to ensure that there are not typos in category and sub-category names (that might be simpler than having a look-aside table).

Reply:
It depends on what that means: is "Information displays" a subcategory of "Layout widgets"  If not, then I don't think it's the same thing.

I think it would be better to just have multiple @Category annotations (assuming 'Information displays' isn't related to 'Layout widgets'):
```dart
/// @Category(['Layout widgets', 'Single-child layout widgets'])
/// @Category(['Information displays'])
```